src/libs/codemirror-mgl.ts

```diff
@@ -1,174 +1,219 @@
+    }
+    return [];
+  };
+}
+
+function createMaplibreStyleLinter() {
+  return (view: EditorView) => {
+    const text = view.state.doc.toString();
+
+    try {
+      // Parse the JSON. The jsonParseLinter will handle pure JSON syntax errors.
+      const parsedJson = JSON.parse(text);
+      const ast = jsonToAst(text);
+
+      // Run the maplibre-gl-style-spec validator.
+      const validationErrors = validateStyleMin(parsedJson);
+      const linterErrors = validationErrors.map(err => {
+        return {
+          key: err.message.split(":")[0],
+          message: err.message,
         };
-        found.push(err);
-      }
-      else if (key) {
-        const path = key.replace(/^\[|\]$/g, "").split(/\.|[[\]]+/).filter(Boolean);
-        const parsedError = getArrayPositionalFromAst(ast, path);
-        if (!parsedError) {
-          console.warn("Something went wrong parsing error:", error);
-          return;
-        }
+      });
+      return getDiagnosticsFromExpressionErrors(linterErrors, ast);
+    } catch {
+      // The built-in JSON linter handles JSON parsing errors, so we don't need to report them again.
+    }
+    return [];
+  };
+}
 
-        const {loc} = parsedError;
-        const {start, end} = loc;
+function createMaplibreExpressionLinter(spec: StylePropertySpecification) {
+  return (view: EditorView) => {
+    const text = view.state.doc.toString();
+    const parsedJson = JSON.parse(text);
+    const ast = jsonToAst(text);
+    const out = expression.createExpression(parsedJson, spec);
+    if (out?.result !== "error") {
+      return [];
+    }
+    const errors = out.value;
+    return getDiagnosticsFromExpressionErrors(errors, ast);
+  };
+}
 
-        found.push({
-          from: CodeMirror.Pos(start.line - 1, start.column),
-          to: CodeMirror.Pos(end.line - 1, end.column),
-          message: message,
-        });
+export function createEditor(props: {
+  parent: HTMLElement,
+  value: string,
+  lintType: LintType,
+  onChange: (value: string) => void,
+  onFocus: () => void,
+  onBlur: () => void,
+  spec?: StylePropertySpecification,
+}): EditorView {
+  let specificLinter: (view: EditorView) => Diagnostic[] = () => [];
+  switch (props.lintType) {
+    case "style":
+      specificLinter = createMaplibreStyleLinter();
+      break;
+    case "layer":
+      specificLinter = createMaplibreLayerLinter();
+      break;
+    case "expression":
+      if (!props.spec) {
+        throw new Error("spec is required for expression mode");
       }
-    });
+      specificLinter = createMaplibreExpressionLinter(props.spec);
+      break;
+    case "json":
+      specificLinter = () => [];
+      break;
   }
 
-  return found;
-});
+  return new EditorView({
+    doc: props.value,
+    extensions: [
+      basicSetup,
+      json(),
+      oneDark,
+      new Compartment().of(EditorState.tabSize.of(2)),
+      EditorView.theme({
+        "&": {
+          fontSize: "9pt"
```

Comment:
This is more of a global issue, but I think we should increase the size of fonts in maputnik.
9pt is a bit small, at least for my eyes.

Reply:
I agree, I mainly wanted to keep the appearance similar to the current implantation. I think a size overhaul is outside the scope of this PR.